[BUG] Authentik: OIDC authentication failed

Hi!

I am trying to set up logging with Authentik following the instructions, but the login fails, showing "OIDC authentication failed," and the logs indicate:

meshmonitor  | [ERROR] OIDC callback error: SqliteError: UNIQUE constraint failed: users.username
meshmonitor  |     at UserModel.create (file:///app/dist/server/models/User.js:40:29)
meshmonitor  |     at handleOIDCCallback (file:///app/dist/server/auth/oidcAuth.js:126:52)
meshmonitor  |     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
meshmonitor  |     at async file:///app/dist/server/routes/authRoutes.js:251:22 {
meshmonitor  |   code: 'SQLITE_CONSTRAINT_UNIQUE'
meshmonitor  | }
meshmonitor  | [ERROR] OIDC callback error: SqliteError: UNIQUE constraint failed: users.username
meshmonitor  |     at UserModel.create (file:///app/dist/server/models/User.js:40:29)
meshmonitor  |     at handleOIDCCallback (file:///app/dist/server/auth/oidcAuth.js:126:52)
meshmonitor  |     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
meshmonitor  |     at async file:///app/dist/server/routes/authRoutes.js:251:22 {
meshmonitor  |   code: 'SQLITE_CONSTRAINT_UNIQUE'
meshmonitor  | }`

--- a/src/server/auth/oidcAuth.ts
+++ b/src/server/auth/oidcAuth.ts
@@ -149,7 +149,7 @@ export async function handleOIDCCallback(
     // Create username from claims
     const username = preferredUsername || email?.split('@')[0] || sub.substring(0, 20);
 
-    // Check if user exists
+    // Check if user exists by OIDC subject
     let user = databaseService.userModel.findByOIDCSubject(sub);
 
     if (user) {
@@ -171,29 +171,60 @@ export async function handleOIDCCallback(
         throw new Error('OIDC user not found and auto-creation is disabled');
       }
 
-      // Create new user
-      user = await databaseService.userModel.create({
-        username,
-        email,
-        displayName: name,
-        authProvider: 'oidc',
-        oidcSubject: sub,
-        isAdmin: false
-      });
+      // Check if a native-login user exists with the same username or email
+      let existingUser = databaseService.userModel.findByUsername(username);
 
-      // Grant default permissions
-      databaseService.permissionModel.grantDefaultPermissions(user.id, false);
+      // If no match by username, try matching by email (if provided)
+      if (!existingUser && email) {
+        existingUser = databaseService.userModel.findByEmail(email);
+      }
 
-      logger.debug(`✅ OIDC user auto-created: ${user.username}`);
+      if (existingUser && existingUser.authProvider === 'local') {
+        // Migrate existing native-login user to OIDC
+        logger.info(`🔄 Migrating existing native-login user '${existingUser.username}' to OIDC`);
 
-      // Audit log
-      databaseService.auditLog(
-        user.id,
-        'oidc_user_created',
-        'users',
-        JSON.stringify({ userId: user.id, username, oidcSubject: sub }),
-        null
-      );
+        user = databaseService.userModel.migrateToOIDC(
+          existingUser.id,
+          sub,
+          email,
+          name
+        )!;
+
+        // Audit log
+        databaseService.auditLog(
+          user.id,
+          'user_migrated_to_oidc',
+          'users',
+          JSON.stringify({ userId: user.id, username: user.username, oidcSubject: sub }),
+          null
+        );
+
+        logger.debug(`✅ User migrated to OIDC: ${user.username}`);
+      } else {
+        // Create new user
+        user = await databaseService.userModel.create({
+          username,
+          email,
+          displayName: name,
+          authProvider: 'oidc',
+          oidcSubject: sub,
+          isAdmin: false
+        });
+
+        // Grant default permissions
+        databaseService.permissionModel.grantDefaultPermissions(user.id, false);
+
+        logger.debug(`✅ OIDC user auto-created: ${user.username}`);
+
+        // Audit log
+        databaseService.auditLog(
+          user.id,
+          'oidc_user_created',
+          'users',
+          JSON.stringify({ userId: user.id, username, oidcSubject: sub }),
+          null
+        );
+      }
     }
 
     return user;

--- a/src/server/models/User.ts
+++ b/src/server/models/User.ts
@@ -235,6 +235,64 @@ export class UserModel {
   }
 
   /**
+   * Migrate a native-login user to OIDC authentication
+   * This preserves all user settings, permissions, and data while
+   * converting the authentication method to OIDC
+   */
+  migrateToOIDC(id: number, oidcSubject: string, email?: string, displayName?: string): User | null {
+    const user = this.findById(id);
+    if (!user) {
+      throw new Error('User not found');
+    }
+
+    if (user.authProvider === 'oidc') {
+      throw new Error('User is already using OIDC authentication');
+    }
+
+    const stmt = this.db.prepare(`
+      UPDATE users
+      SET
+        auth_provider = 'oidc',
+        oidc_subject = ?,
+        password_hash = NULL,
+        email = COALESCE(?, email),
+        display_name = COALESCE(?, display_name),
+        last_login_at = ?
+      WHERE id = ?
+    `);
+
+    stmt.run(
+      oidcSubject,
+      email || null,
+      displayName || null,
+      Date.now(),
+      id
+    );
+
+    return this.findById(id);
+  }
+
+  /**
+   * Find user by email (case-insensitive)
+   */
+  findByEmail(email: string): User | null {
+    const stmt = this.db.prepare(`
+      SELECT
+        id, username, password_hash as passwordHash, email, display_name as displayName,
+        auth_provider as authProvider, oidc_subject as oidcSubject,
+        is_admin as isAdmin, is_active as isActive,
+        created_at as createdAt, last_login_at as lastLoginAt, created_by as createdBy
+      FROM users
+      WHERE LOWER(email) = LOWER(?)
+    `);
+
+    const row = stmt.get(email) as any;
+    if (!row) return null;
+
+    return this.mapRowToUser(row);
+  }
+
+  /**
    * Delete (deactivate) a user
    */
   delete(id: number): void {

--- a/src/server/routes/authRoutes.test.ts
+++ b/src/server/routes/authRoutes.test.ts
@@ -603,4 +603,192 @@ describe('Authentication Routes', () => {
       expect(typeof response.body.localAuthDisabled).toBe('boolean');
     });
   });
+
+  describe('OIDC Account Migration', () => {
+    it('should migrate native-login user to OIDC on first OIDC login', async () => {
+      // Create a native-login user
+      const nativeUser = await userModel.create({
+        username: 'migrateuser',
+        password: 'password123',
+        email: 'migrate@example.com',
+        authProvider: 'local',
+        isAdmin: false
+      });
+
+      // Grant permissions
+      permissionModel.grantDefaultPermissions(nativeUser.id, false);
+
+      // Verify the user exists as a native-login user
+      let user = userModel.findById(nativeUser.id);
+      expect(user).toBeTruthy();
+      expect(user!.authProvider).toBe('local');
+      expect(user!.passwordHash).toBeTruthy();
+
+      // Simulate OIDC migration by directly calling migrateToOIDC
+      const oidcSubject = 'oidc-sub-123';
+      const migratedUser = userModel.migrateToOIDC(
+        nativeUser.id,
+        oidcSubject,
+        'migrate@example.com',
+        'Migrate User'
+      );
+
+      // Verify migration
+      expect(migratedUser).toBeTruthy();
+      expect(migratedUser!.id).toBe(nativeUser.id); // Same user ID
+      expect(migratedUser!.username).toBe('migrateuser'); // Same username
+      expect(migratedUser!.authProvider).toBe('oidc');
+      expect(migratedUser!.oidcSubject).toBe(oidcSubject);
+      expect(migratedUser!.passwordHash).toBeNull(); // Password hash removed
+      expect(migratedUser!.email).toBe('migrate@example.com');
+      expect(migratedUser!.displayName).toBe('Migrate User');
+
+      // Verify old password no longer works
+      const oldAuth = await userModel.authenticate('migrateuser', 'password123');
+      expect(oldAuth).toBeNull();
+    });
+
+    it('should preserve user permissions when migrating to OIDC', async () => {
+      // Create a native-login user
+      const nativeUser = await userModel.create({
+        username: 'permissionuser',
+        password: 'password123',
+        email: 'permissions@example.com',
+        authProvider: 'local',
+        isAdmin: true
+      });
+
+      // Grant specific permissions
+      permissionModel.grantDefaultPermissions(nativeUser.id, true);
+
+      // Get permissions before migration
+      const permissionsBefore = permissionModel.getUserPermissions(nativeUser.id);
+
+      // Migrate to OIDC
+      const migratedUser = userModel.migrateToOIDC(
+        nativeUser.id,
+        'oidc-sub-456',
+        'permissions@example.com',
+        'Permission User'
+      );
+
+      // Get permissions after migration
+      const permissionsAfter = permissionModel.getUserPermissions(migratedUser!.id);
+
+      // Verify permissions are preserved
+      expect(permissionsAfter).toEqual(permissionsBefore);
+
+      // Verify admin status is preserved
+      expect(migratedUser!.isAdmin).toBe(true);
+    });
+
+    it('should find user by email for migration when username differs', async () => {
+      // Create a native-login user
+      const nativeUser = await userModel.create({
+        username: 'oldusername',
+        password: 'password123',
+        email: 'email-match@example.com',
+        authProvider: 'local',
+        isAdmin: false
+      });
+
+      // Verify findByEmail works (case-insensitive)
+      const foundUser = userModel.findByEmail('EMAIL-match@example.com');
+      expect(foundUser).toBeTruthy();
+      expect(foundUser!.id).toBe(nativeUser.id);
+      expect(foundUser!.username).toBe('oldusername');
+    });
+
+    it('should prevent migrating an already-OIDC user', async () => {
+      // Create an OIDC user
+      const oidcUser = await userModel.create({
+        username: 'oidcuser',
+        email: 'oidc@example.com',
+        authProvider: 'oidc',
+        oidcSubject: 'oidc-sub-789',
+        isAdmin: false
+      });
+
+      // Try to migrate again
+      expect(() => {
+        userModel.migrateToOIDC(
+          oidcUser.id,
+          'oidc-sub-new',
+          'oidc@example.com',
+          'OIDC User'
+        );
+      }).toThrow('User is already using OIDC authentication');
+    });
+
+    it('should update last login timestamp during migration', async () => {
+      // Create a native-login user
+      const nativeUser = await userModel.create({
+        username: 'timestampuser',
+        password: 'password123',
+        email: 'timestamp@example.com',
+        authProvider: 'local',
+        isAdmin: false
+      });
+
+      const beforeTimestamp = Date.now();
+
+      // Migrate to OIDC
+      const migratedUser = userModel.migrateToOIDC(
+        nativeUser.id,
+        'oidc-sub-timestamp',
+        'timestamp@example.com',
+        'Timestamp User'
+      );
+
+      // Verify last login was updated
+      expect(migratedUser!.lastLoginAt).toBeTruthy();
+      expect(migratedUser!.lastLoginAt!).toBeGreaterThanOrEqual(beforeTimestamp);
+    });
+
+    it('should preserve email and display name when not provided during migration', async () => {
+      // Create a native-login user with existing data
+      const nativeUser = await userModel.create({
+        username: 'preserveuser',
+        password: 'password123',
+        email: 'preserve@example.com',
+        displayName: 'Original Name',
+        authProvider: 'local',
+        isAdmin: false
+      });
+
+      // Migrate without providing email/displayName
+      const migratedUser = userModel.migrateToOIDC(
+        nativeUser.id,
+        'oidc-sub-preserve'
+      );
+
+      // Verify original values are preserved
+      expect(migratedUser!.email).toBe('preserve@example.com');
+      expect(migratedUser!.displayName).toBe('Original Name');
+    });
+
+    it('should update email and display name when provided during migration', async () => {
+      // Create a native-login user with existing data
+      const nativeUser = await userModel.create({
+        username: 'updateuser',
+        password: 'password123',
+        email: 'old@example.com',
+        displayName: 'Old Name',
+        authProvider: 'local',
+        isAdmin: false
+      });
+
+      // Migrate with new email/displayName
+      const migratedUser = userModel.migrateToOIDC(
+        nativeUser.id,
+        'oidc-sub-update',
+        'new@example.com',
+        'New Name'
+      );
+
+      // Verify values were updated
+      expect(migratedUser!.email).toBe('new@example.com');
+      expect(migratedUser!.displayName).toBe('New Name');
+    });
+  });
 });

